client/tinychain/ml/linalg.py

```diff
@@ -274,3 +277,81 @@ def svd(matrix: Tensor) -> Tuple:
     """Return the singular value decomposition of the given `matrix`."""
 
     raise NotImplementedError("singular value decomposition")
+
+
+def householder_bidiag(U: Tensor, W: Tensor, e: Tensor):
+    """Householder's reduction to bidiagonal form"""
+    @post_op
+    def check_rows_num(cxt, U: Tensor, scale: F32, i: UInt, m: UInt, n: UInt, l: UInt):
+        scale = F32(tc.linalg.norm(U[i:, i]))**2
+
+        # @post_op
+        def change_rows(cxt, U: Tensor, scale: F32, i: UInt, l: UInt, n: UInt):
+            s = F32(After(U[i:, i].write(U[i:, i] / scale),
+                    F32(tc.linalg.norm(U[i:, i]))))
+            f = F32(U[i, i])
+            g = F32(If(f >= F32(0), -s, s))*1
```

Comment:
rather than multiply by one it's more efficient to assign an explicit name within the op context:
`cxt.g = F32(If(f >= 0, -s, s))`